crates/runtime/src/accelerated_table/refresh_task/changes.rs

```diff
@@ -0,0 +1,247 @@
+
+                            if let Some(cache_provider) = &cache_provider {
+                                if let Err(e) = cache_provider
+                                    .invalidate_for_table(&dataset_name.to_string())
+                                    .await
+                                {
+                                    tracing::error!(
+                                        "Failed to invalidate cached results for dataset {}: {e}",
+                                        &dataset_name.to_string()
+                                    );
+                                }
+                            }
+                        }
+                        Err(e) => {
+                            self.mark_dataset_status(status::ComponentStatus::Error)
+                                .await;
+                            tracing::error!("Error writing change for {dataset_name}: {e}");
+                        }
+                    }
+                }
+                Err(e) => {
+                    tracing::error!("Changes stream error for {dataset_name}: {e}");
+                    self.mark_dataset_status(status::ComponentStatus::Error)
+                        .await;
+                }
+            }
+        }
+
+        tracing::warn!("Changes stream ended for dataset {dataset_name}");
+
+        Ok(())
+    }
+
+    async fn write_change(
+        &self,
+        change_batch: ChangeBatch,
+    ) -> crate::accelerated_table::Result<()> {
+        let dataset_name = self.dataset_name.clone();
+        let deletion_provider = get_deletion_provider(Arc::clone(&self.accelerator))
+            .context(crate::accelerated_table::AcceleratedTableDoesntSupportDeleteSnafu)?;
+
+        for row in 0..change_batch.record.num_rows() {
+            let op = change_batch.op(row);
+            match op {
+                "d" => {
+                    let inner_data: RecordBatch = change_batch.data(row);
+                    let primary_keys = change_batch.primary_keys(row);
+                    let primary_key_log_fmt =
+                        Self::get_primary_key_log_fmt(&inner_data, &primary_keys)?;
+                    let delete_where_exprs =
+                        Self::get_delete_where_expr(&inner_data, primary_keys)?;
+
+                    tracing::info!("Deleting data for {dataset_name} where {primary_key_log_fmt}");
+
+                    let ctx = SessionContext::new();
+                    let session_state = ctx.state();
+
+                    let delete_plan = deletion_provider
+                        .delete_from(&session_state, &delete_where_exprs)
+                        .await
+                        .context(crate::accelerated_table::FailedToWriteDataSnafu)?;
+
+                    collect(delete_plan, ctx.task_ctx())
+                        .await
+                        .context(crate::accelerated_table::FailedToWriteDataSnafu)?;
+                }
+                "c" | "u" | "r" => {
```

Comment:
Why `read` event handled as an upsertion?

Reply:
`read` is part of the initial snapshot that Debezium ingests into Kafka. https://debezium.io/documentation/reference/2.7/connectors/postgresql.html#postgresql-snapshots